src/Symfony/Component/PropertyInfo/Type.php

```diff
@@ -69,8 +72,31 @@ public function __construct(string $builtinType, bool $nullable = false, string
         $this->nullable = $nullable;
         $this->class = $class;
         $this->collection = $collection;
-        $this->collectionKeyType = $collectionKeyType;
-        $this->collectionValueType = $collectionValueType;
+        $this->collectionKeyType = $this->validateCollectionArgument($collectionKeyType, 5, '$collectionKeyType');
+        $this->collectionValueType = $this->validateCollectionArgument($collectionValueType, 6, '$collectionValueType');
+    }
+
+    private function validateCollectionArgument($collectionArgument, int $argumentIndex, string $argumentName): ?array
+    {
+        if (null === $collectionArgument) {
+            return null;
+        }
+
+        if (null !== $collectionArgument && !\is_array($collectionArgument) && !$collectionArgument instanceof self) {
+            throw new \TypeError(sprintf('"%s()": Argument #%d (%s) must be of type "array", "%s" or "null", "%s" given.', __METHOD__, $argumentIndex, $argumentName, self::class, get_debug_type($collectionArgument)));
+        }
+
+        if (\is_array($collectionArgument)) {
+            foreach ($collectionArgument as $type) {
+                if (!$type instanceof self) {
+                    throw new \TypeError(sprintf('"%s()": Argument #%d (%s) must be an array with items of type "%s", "%s" given.', __METHOD__, $argumentIndex, $argumentName, self::class, get_debug_type($collectionArgument)));
```

Comment:
that's wrong, it must not "be an array with items of type %s", null and instance of %s are valide too